monai/transforms/spatial/array.py

```diff
@@ -2637,7 +2637,9 @@ class GridPatch(Transform):
             which are, respectively, the minimum and maximum of the sum of intensities of a patch across all dimensions
             and channels. Also "random" creates a random order of patches.
             By default no sorting is being done and patches are returned in a row-major order.
-        pad_mode: refer to  NumpyPadMode and PytorchPadMode. Defaults to ``"constant"``.
+        threshold: a value to keep only the patches whose sum of intensities are less than the threshold.
```

Comment:
Maybe we can change this `threshold` arg to `filter_fn` directly? And default to `None`?
It's more flexible and the code logic can be easier, refer to:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/croppad/array.py#L691

What do you think?

Thanks.

Reply:
This is how I implemented in previous commits but the problem is that here we need to pass the threshold to the function, which will cause problem in distributed trainings since it need to be pickled and it cannot be a lambda function or a function defined locally in the worker. Then it will create a hassle for the user on where and how to create such filter function. So it would better to make it simple and use threshold for its most common use case.

Reply:
Hi @drbeh ,

I think it's not a problem to pass a function as the arg in multi-gpu program, many MONAI APIs do that, and I remember it only affects the `spawn` multi-processing?
To support pickle, you can just define the function and pass it as the arg.
The benefit of using `filter_fn` is that it can simplify the code logic and support more kinds of filters in the future.
Please feel free to correct me if I misunderstood the logic.
@ericspod @wyli @rijobro What do you think?

Thanks.

Reply:
@Nic-Ma I know that, and if you check the previous commits of this PR, you can see the exact implementation that you say. But consider the MIL pipeline that we have: https://github.com/Project-MONAI/tutorials/blob/main/pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py

It uses `spawn` and the way that the arguments are passed, you cannot create such filtering function based on the input args. I know that we can change the pipeline but the point is users need to know too much for that too work. It's a simple and common functionality to threshold and I don't think that we should make it more difficult for the user to use it.

Reply:
One of the issues here is that arguments get pickled when passed to a spawned process and a pickle of a function only contains a reference to it's name and not the bytecode of the function itself. If a function defined with a name that can be imported is passed to a spawned process that process will import the function and get the reference to it, but a function defined locally (eg. in a notebook) cannot be imported and will cause an error. So this sometimes works and sometimes doesn't for non-obvious reasons, so it's probably best to use the threshold value instead of a callable. We've seen this before under Windows with some transforms working inside dataloaders with multiple workers sometimes and sometimes not.

Reply:
OK, let's keep this implementation for now, thanks for the discussion.